testplan/web_ui/testing/src/index.js

```diff
@@ -4,29 +4,45 @@ import BatchReport from './Report/BatchReport';
 import InteractiveReport from './Report/InteractiveReport';
 import EmptyReport from './Report/EmptyReport';
 import {POLL_MS} from './Common/defaults.js';
+import Home from './Common/Home';
 
 // import registerServiceWorker from './registerServiceWorker';
 import 'bootstrap/dist/css/bootstrap.min.css';
-import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
+import { BrowserRouter, Route, Switch, Redirect } from "react-router-dom";
 
 /**
  * This single App provides multiple functions controlled via the URL path
  * accessed. We are using React-Router to control which type of report is
  * rendered and to extract the report UID from the URL when necessary.
+ *
+ * We use "strict" here to ensure that paths end with a trailing slash. This
+ * makes the UI-rendered hash routes more distinguishable from the
```

Comment:
I don't quite understand this comment - can you explain it to me in more detail? In particular I don't understand what you mean by "UI-rendered hash routes" and "server-rendered regular routes".

Also the comment doesn't quite seem to match with the code below - the redirect route is the first case in the switch not the last one?

Reply:
Ah yeah this made more sense before I split it into its own file. And the "UI-rendered has routes" doesn't make sense without seeing the new BatchReport. 

I'll fix the description but in short, we'll have this URL:

`/testplan/path1/xyz/?a=1&b=2#/MyTestPlan/My%20Suite/my_testcase?filter=passing`

`/testplan/path1/xyz/?a=1&b=2` <- this bit is sent over the net to our API
`/MyTestPlan/My%20Suite/my_testcase?filter=passing` <- this bit controls / reflects the UI state and is not sent over the net

Reply:
So you're suggesting we store the navigation state as a # fragment? I wasn't aware that was the design. Is there any reason we can't just set the URL to something like `/reports/<UID>/My%20Suite/my_testcase?filter=passing`? Seems simpler to me.

Reply:
Seems to me this would complicate things since the browser will navigate to the portion before the hash. Try it: `window.location.pathname = '/Morgan-Stanley/testplan/pull/390'`.

This is do-able but we'd need to keep the API in sync with the routes used in the UI so it knows to ignore some, plus the roundtrip overhead of the web request.

So, simpler how?

Reply:
It's just that is how the examples I've seen in the react-router work e.g. https://reacttraining.com/react-router/web/example/url-params , e.g. you click the "Netflix" link and it redirects to /netflix. I don't think it actually reloads the page in the browser, it's just JS manipulations.

Generally I've only seen # marks in URLs used to jump to a specific section on a page, not to redirect to a new page entirely. Is there a specific example / documention you've read which suggests to do the routing using # fragments?

Reply:
This is getting a disproportionate amount of attention. https://reacttraining.com/react-router/web/guides/primary-components

Reply:
Thanks for the link, I wasn't aware of the HashRouter component, I've only ever used the BrowserRouter before.

So, to my undestanding your design is to continue using the BrowserRouter at the top level to distinguish between the batch and interactive modes, and to possible add some common home page or other pages (a search page?) in future. But then the HashRouter will be used to distinguish between navigation states within each UI. Is my understanding correct? This seems like an important design decision to me so I just want to make sure we get it right.

Reply:
On the BatchReport.js side at least, yes, using HashRouter for UI-only navigation seems the way to go. I think in the end we _may_ want to the same for the InteractiveReport.js side as well, though its benefits there are less certain.

For example with the BatchReport you'll be able to get your UI into a certain state (which right now just consists of what level of the report you're at, but could conceivably include user params like hide / show passing), then you or someone else could revisit that URL and restore that UI state, which I see this as the biggest benefit. The InteractiveReport needs the local server running and can't assume the server's state hasn't changed, so a link ( == testplan report + its UI state) can't be revisited in the same way. It would make the forward-back buttons work though.